stock_storage_type/models/stock_quant.py

```diff
@@ -0,0 +1,156 @@
+# Copyright 2020 Camptocamp SA
+# License AGPL-3.0 or later (https://www.gnu.org/licenses/agpl)
+from odoo import _, models, api
+from odoo.exceptions import ValidationError
+
+
+class StockQuant(models.Model):
+
+    _inherit = 'stock.quant'
+
+    @api.constrains('package_id', 'location_id', 'lot_id', 'product_id')
+    def _check_storage_type(self):
+        """
+        Check if at least one location storage type allows the package storage
+        type into the quant's location
+        """
+        for quant in self:
+            location = quant.location_id
+            pack_storage_type = quant.package_id.package_storage_type_id
+            loc_storage_types = location.allowed_location_storage_type_ids
+            if (
+                not quant.package_id
+                or not pack_storage_type
+                or not loc_storage_types
+            ):
+                continue
+            lst_allowed_for_pst = loc_storage_types.filtered(
+                lambda lst: pack_storage_type in lst.package_storage_type_ids
+            )
+            if not lst_allowed_for_pst:
+                raise ValidationError(
+                    _(
+                        "Package storage type %s is not allowed into "
+                        "Location %s"
+                    ) % (pack_storage_type.name, location.name)
+                )
+            allowed = False
+            package_quants = quant.package_id.mapped('quant_ids')
+            package_products = package_quants.mapped('product_id')
+            package_lots = package_quants.mapped('lot_id')
+            other_quants_in_location = self.search(
+                [
+                    ('location_id', '=', location.id),
+                    ('id', 'not in', package_quants.ids)
```

Comment:
We must exclude all previous quants that are not there anymore, i.e. quantity = 0

```suggestion
                    ('id', 'not in', package_quants.ids),
                    ('quantity', '!=', 0)
```